File: server/rtt.py
"""Steady-stream RTT monitoring (mirrored in web/rtt.js).

One side sends PING packets at a fixed interval, each carrying the reading of
its own clock at hand-off. The other side turns each one around immediately as
a PONG, stamping its own arrival time, so a single exchange yields

    rtt      = pong arrival - ping send        (one clock, no sync needed)
    up leg   = echo_recv_ts - send_ts          (+ the clock offset)
    down leg = pong arrival - echo_recv_ts     (- the clock offset)

The offset between the two clocks is unknown but (over a run) constant, so the
legs mean nothing in absolute terms - only against their own minimum.
`up_excess = up - min(up)` is the delay the forward direction added beyond its
best case in this run, and likewise for the reverse. That split is the point of
the exercise: a rising `up_excess` with a flat `down_excess` says the queue is
on the browser's uplink (very often the browser's own send queue), which is
exactly the ambiguity a plain RTT number leaves open.

Two halves, both used on both ends:

* `RttMonitor` - the initiator: matches PONGs to PINGs, keeps srtt/min-RTT and
  per-sample records for plotting.
* `RttStreamReceiver` - the responder: parses the timestamps *inside* the
  stream it receives. It never needs the reply, so it works even when the
  return path is broken: arrival spacing vs send spacing, one-way delay excess,
  RFC 3550 jitter, and loss/reordering from the sequence numbers.

The stream is deliberately thin (a 64-byte packet every 50 ms is 10 kbps) so it
measures the path's standing delay rather than creating queueing of its own.
"""

import asyncio
from typing import Any, Callable, Dict, List, Optional

import protocol as proto
from trains import _percentiles

# RFC 6298 smoothing, same constants as transport_stats.SenderCore.
ALPHA = 1 / 8
BETA = 1 / 4
# RFC 3550 interarrival jitter gain.
JITTER_GAIN = 1 / 16
BIN_MS = 100.0


def _bin_timeline(rows: List[Dict[str, Any]], start: float, keys: List[str]) -> List[Dict[str, Any]]:
    """Group per-packet rows into BIN_MS bins, median per key (for plots)."""
    bins: Dict[int, List[Dict[str, Any]]] = {}
    for r in rows:
        bins.setdefault(int((r["t"] - start) // BIN_MS), []).append(r)
    out = []
    for idx in sorted(bins):
        group = bins[idx]
        row: Dict[str, Any] = {"t": idx * BIN_MS, "packets": len(group)}
        for key in keys:
            vals = [g[key] for g in group if g.get(key) is not None]
            row[key] = _percentiles(vals)["p50"] if vals else None
        out.append(row)
    return out


class RttMonitor:
    """Initiator side: sends PINGs, matches the PONGs that come back."""

    def __init__(self, keep_records: bool = True) -> None:
        self.sent = 0
        self.replies = 0
        self.duplicates = 0
        self.reordered = 0
        self.pending: Dict[int, float] = {}  # seq -> our send time
        self.srtt: Optional[float] = None
        self.rttvar: Optional[float] = None
        self.min_rtt: Optional[float] = None
        self.min_up: Optional[float] = None
        self.min_down: Optional[float] = None
        self.max_seq = -1
        self.records: Optional[List[Dict[str, Any]]] = [] if keep_records else None
        self._rtts: List[float] = []
        self._ups: List[float] = []
        self._downs: List[float] = []
        self._ipdv: List[float] = []
        self._prev_rtt: Optional[float] = None
        self.first_send: Optional[float] = None

    def on_send(self, seq: int, t: float) -> None:
        self.sent += 1
        self.pending[seq] = t
        if self.first_send is None:
            self.first_send = t

    def on_pong(self, pkt: proto.Pong, t: float) -> Optional[Dict[str, Any]]:
        if self.pending.pop(pkt.seq, None) is None and pkt.seq <= self.max_seq:
            self.duplicates += 1  # already accounted for (or never ours)
            return None
        # RTT comes from the echoed timestamp, so it survives a lost record.
        rtt = t - pkt.send_ts
        self.replies += 1
        if pkt.seq < self.max_seq:
            self.reordered += 1
        self.max_seq = max(self.max_seq, pkt.seq)

        if self.srtt is None:
            self.srtt, self.rttvar = rtt, rtt / 2
        else:
            self.rttvar = (1 - BETA) * self.rttvar + BETA * abs(self.srtt - rtt)
            self.srtt = (1 - ALPHA) * self.srtt + ALPHA * rtt
        self.min_rtt = rtt if self.min_rtt is None else min(self.min_rtt, rtt)
        if self._prev_rtt is not None:
            self._ipdv.append(rtt - self._prev_rtt)
        self._prev_rtt = rtt

        up = pkt.echo_recv_ts - pkt.send_ts  # + clock offset
        down = t - pkt.echo_recv_ts  # - clock offset
        self.min_up = up if self.min_up is None else min(self.min_up, up)
        self.min_down = down if self.min_down is None else min(self.min_down, down)
        self._rtts.append(rtt)
        self._ups.append(up)
        self._downs.append(down)

        rec = {
            "seq": pkt.seq,
            "t": pkt.send_ts,
            "rtt_ms": rtt,
            "queue_ms": rtt - self.min_rtt,
            "up_ms": up,
            "down_ms": down,
            "srtt_ms": self.srtt,
        }
        if self.records is not None:
            self.records.append(rec)
        return rec

    def outstanding(self, t: float, timeout_ms: float) -> int:
        """PINGs with no reply yet, older than `timeout_ms` (i.e. lost)."""
        return sum(1 for ts in self.pending.values() if t - ts > timeout_ms)

    def summary(self, keep_records: bool = False) -> Optional[Dict[str, Any]]:
        if not self.replies:
            return None
        min_rtt = self.min_rtt or 0.0
        rows = self.records or []
        out: Dict[str, Any] = {
            "sent": self.sent,
            "replies": self.replies,
            "lost": max(0, self.sent - self.replies),
            "loss_pct": 100.0 * max(0, self.sent - self.replies) / self.sent if self.sent else None,
            "reordered": self.reordered,
            "duplicates": self.duplicates,
            "min_rtt_ms": self.min_rtt,
            "srtt_ms": self.srtt,
            "rttvar_ms": self.rttvar,
            "rtt_ms": _percentiles(self._rtts),
            # standing queue: how far above the best RTT of the run we sat
            "queue_ms": _percentiles([r - min_rtt for r in self._rtts]),
            # |consecutive RTT change|: short-term variation, offset-free
            "ipdv_abs_ms": _percentiles([abs(d) for d in self._ipdv]),
            # each leg relative to its own minimum (the clock offset cancels)
            "up_excess_ms": _percentiles([u - self.min_up for u in self._ups]) if self.min_up is not None else None,
            "down_excess_ms": _percentiles([d - self.min_down for d in self._downs]) if self.min_down is not None else None,
            "timeline": _bin_timeline(rows, self.first_send or 0.0, ["rtt_ms", "queue_ms", "srtt_ms"]),
        }
        if keep_records:
            out["records"] = rows
        return out


class RttStreamReceiver:
    """Responder side: what the timestamps *in* the stream say, reply aside."""

    def __init__(self, keep_records: bool = True) -> None:
        self.received = 0
        self.duplicates = 0
        self.reordered = 0
        self.first_seq: Optional[int] = None
        self.max_seq = -1
        self.jitter = 0.0  # RFC 3550 interarrival jitter, ms
        self.min_owd: Optional[float] = None
        self.records: Optional[List[Dict[str, Any]]] = [] if keep_records else None
        self._seen: set = set()
        self._prev: Optional[tuple] = None  # (send_ts, recv_ts) of the last arrival
        self._owds: List[float] = []
        self._iat: List[float] = []
        self._send_iat: List[float] = []
        self._ipdv: List[float] = []
        self.first_recv: Optional[float] = None

    def on_ping(self, pkt: proto.Ping, recv_ts: float) -> Dict[str, Any]:
        if pkt.seq in self._seen:
            self.duplicates += 1
        self._seen.add(pkt.seq)
        self.received += 1
        if self.first_seq is None:
            self.first_seq = pkt.seq
            self.first_recv = recv_ts
        if pkt.seq < self.max_seq:
            self.reordered += 1
        self.max_seq = max(self.max_seq, pkt.seq)

        owd = recv_ts - pkt.send_ts  # one-way delay + clock offset
        self.min_owd = owd if self.min_owd is None else min(self.min_owd, owd)
        iat = send_iat = d = None
        if self._prev is not None:
            send_iat = pkt.send_ts - self._prev[0]
            iat = recv_ts - self._prev[1]
            d = iat - send_iat  # RFC 3550 D(i-1, i): spacing change in flight
            self.jitter += (abs(d) - self.jitter) * JITTER_GAIN
            self._iat.append(iat)
            self._send_iat.append(send_iat)
            self._ipdv.append(d)
        self._prev = (pkt.send_ts, recv_ts)
        self._owds.append(owd)

        rec = {
            "seq": pkt.seq,
            "t": recv_ts,
            "send_ts": pkt.send_ts,
            "owd_ms": owd,
            "owd_excess_ms": owd - self.min_owd,
            "iat_ms": iat,
            "send_iat_ms": send_iat,
            "ipdv_ms": d,
            "jitter_ms": self.jitter,
        }
        if self.records is not None:
            self.records.append(rec)
        return rec

    def summary(self, keep_records: bool = False) -> Optional[Dict[str, Any]]:
        if not self.received:
            return None
        expected = self.max_seq - (self.first_seq or 0) + 1
        lost = max(0, expected - len(self._seen))
        min_owd = self.min_owd or 0.0
        out: Dict[str, Any] = {
            "received": self.received,
            "expected": expected,
            "lost": lost,
            "loss_pct": 100.0 * lost / expected if expected else None,
            "reordered": self.reordered,
            "duplicates": self.duplicates,
            "jitter_ms": self.jitter,
            # arrival spacing vs the spacing the sender stamped into the packets:
            # if send_iat is steady and iat is not, the delay happened in flight
            # (or in the sender's own queue, which is downstream of the stamp).
            "iat_ms": _percentiles(self._iat),
            "send_iat_ms": _percentiles(self._send_iat),
            "ipdv_abs_ms": _percentiles([abs(d) for d in self._ipdv]),
            # one-way delay above the minimum seen: the queue this direction built
            "owd_excess_ms": _percentiles([o - min_owd for o in self._owds]),
            "timeline": _bin_timeline(
                self.records or [], self.first_recv or 0.0, ["owd_excess_ms", "iat_ms", "jitter_ms"]
            ),
        }
        if keep_records:
            out["records"] = self.records
        return out


async def run_rtt_stream(
    send: Callable[[bytes], None],
    flow: int,
    now_ms: Callable[[], float],
    monitor: RttMonitor,
    interval_ms: float,
    duration_s: float,
    size: int,
    is_open: Callable[[], bool] = lambda: True,
) -> Dict[str, Any]:
    """Send PINGs on a fixed schedule until `duration_s` is up.

    The schedule is absolute (start + n * interval) rather than sleep-per-packet,
    so a late wake-up doesn't push the whole stream out of step.
    """
    start = now_ms()
    end = start + duration_s * 1000
    n = 0
    while is_open():
        t = now_ms()
        if t >= end:
            break
        monitor.on_send(n, t)
        send(proto.encode_ping(flow, n, t, size))
        n += 1
        await asyncio.sleep(max(0.0, (start + n * interval_ms - now_ms()) / 1000))
    return {"sent": n, "interval_ms": interval_ms, "size": size}

File: web/rtt.js
// Steady-stream RTT monitoring - the browser half of server/rtt.py.
//
// One side sends PING packets at a fixed interval, each stamped with its own
// clock at hand-off; the other turns each one around immediately, stamping its
// arrival time. So one exchange gives the round trip and, separately, each leg:
//
//   rtt      = pong arrival - ping send
//   up leg   = echo_recv_ts - send_ts      (+ clock offset)
//   down leg = pong arrival - echo_recv_ts (- clock offset)
//
// The offset is unknown but constant, so the legs only mean something relative
// to their own minimum: up_excess rising while down_excess stays flat says the
// queue is on the uplink. See server/rtt.py for the full reasoning; the classes
// here are line-for-line equivalents so both ends report the same fields.

import * as proto from './protocol.js';

const ALPHA = 1 / 8;
const BETA = 1 / 4;
const JITTER_GAIN = 1 / 16;
const BIN_MS = 100;

export function percentiles(xs) {
  const s = xs.filter((x) => Number.isFinite(x)).sort((a, b) => a - b);
  if (!s.length) return null;
  const p = (q) => s[Math.floor(q * (s.length - 1))];
  return { min: s[0], p25: p(0.25), p50: p(0.5), p75: p(0.75), p95: p(0.95), max: s[s.length - 1],
    mean: s.reduce((a, b) => a + b, 0) / s.length };
}

function binTimeline(rows, start, keys) {
  const bins = new Map();
  for (const r of rows) {
    const idx = Math.floor((r.t - start) / BIN_MS);
    if (!bins.has(idx)) bins.set(idx, []);
    bins.get(idx).push(r);
  }
  return [...bins.keys()].sort((a, b) => a - b).map((idx) => {
    const group = bins.get(idx);
    const row = { t: idx * BIN_MS, packets: group.length };
    for (const k of keys) row[k] = percentiles(group.map((g) => g[k]).filter((v) => v != null))?.p50 ?? null;
    return row;
  });
}

// ---- initiator: sends PINGs, matches the PONGs ------------------------------

export class RttMonitor {
  constructor(keepRecords = true) {
    this.sent = 0; this.replies = 0; this.duplicates = 0; this.reordered = 0;
    this.pending = new Map();           // seq -> our send time
    this.srtt = null; this.rttvar = null; this.minRtt = null;
    this.minUp = null; this.minDown = null;
    this.maxSeq = -1;
    this.records = keepRecords ? [] : null;
    this.rtts = []; this.ups = []; this.downs = []; this.ipdv = [];
    this.prevRtt = null; this.firstSend = null;
    // Send side, to catch what the browser does to the stream *after* we hand
    // it over: the gap we actually achieved between hand-offs, how long the
    // transport took to accept each write, and how deep its own queue got.
    this.sendRecords = [];
    this.sendBySeq = new Map();
    this.sendGaps = []; this.writeMs = []; this.queueDepth = [];
    this.prevSend = null;
  }

  onSend(seq, t) {
    this.sent++;
    this.pending.set(seq, t);
    if (this.firstSend == null) this.firstSend = t;
    const gap = this.prevSend == null ? null : t - this.prevSend;
    this.prevSend = t;
    if (gap != null) this.sendGaps.push(gap);
    const rec = { seq, t, gap_ms: gap, write_ms: null, queue: null };
    this.sendRecords.push(rec);
    this.sendBySeq.set(seq, rec);
    return rec;
  }

  // `writeMs` is the delay before the transport accepted the write (null where
  // the API is synchronous, as on a data channel); `queue` is its own backlog
  // at hand-off - writer.desiredSize for WebTransport, bufferedAmount for RTC.
  onWriteDone(seq, writeMs, queue = null) {
    const rec = this.sendBySeq.get(seq);
    if (!rec) return;
    if (writeMs != null) { rec.write_ms = writeMs; this.writeMs.push(writeMs); }
    if (queue != null) { rec.queue = queue; this.queueDepth.push(queue); }
  }

  onPong(pkt, t) {
    const had = this.pending.delete(pkt.seq);
    if (!had && pkt.seq <= this.maxSeq) { this.duplicates++; return null; }
    const rtt = t - pkt.sendTs;         // echoed timestamp: no lookup needed
    this.replies++;
    if (pkt.seq < this.maxSeq) this.reordered++;
    this.maxSeq = Math.max(this.maxSeq, pkt.seq);

    if (this.srtt == null) { this.srtt = rtt; this.rttvar = rtt / 2; }
    else {
      this.rttvar = (1 - BETA) * this.rttvar + BETA * Math.abs(this.srtt - rtt);
      this.srtt = (1 - ALPHA) * this.srtt + ALPHA * rtt;
    }
    this.minRtt = this.minRtt == null ? rtt : Math.min(this.minRtt, rtt);
    if (this.prevRtt != null) this.ipdv.push(rtt - this.prevRtt);
    this.prevRtt = rtt;

    const up = pkt.echoRecvTs - pkt.sendTs;
    const down = t - pkt.echoRecvTs;
    this.minUp = this.minUp == null ? up : Math.min(this.minUp, up);
    this.minDown = this.minDown == null ? down : Math.min(this.minDown, down);
    this.rtts.push(rtt); this.ups.push(up); this.downs.push(down);

    const rec = {
      seq: pkt.seq, t: pkt.sendTs, rtt_ms: rtt, queue_ms: rtt - this.minRtt,
      up_ms: up, down_ms: down, srtt_ms: this.srtt,
      up_excess_ms: up - this.minUp, down_excess_ms: down - this.minDown,
    };
    this.records?.push(rec);
    return rec;
  }

  outstanding(t, timeoutMs) {
    let n = 0;
    for (const ts of this.pending.values()) if (t - ts > timeoutMs) n++;
    return n;
  }

  summary(keepRecords = false) {
    if (!this.replies) return null;
    const minRtt = this.minRtt ?? 0;
    const lost = Math.max(0, this.sent - this.replies);
    const out = {
      sent: this.sent, replies: this.replies, lost,
      loss_pct: this.sent ? (100 * lost) / this.sent : null,
      reordered: this.reordered, duplicates: this.duplicates,
      min_rtt_ms: this.minRtt, srtt_ms: this.srtt, rttvar_ms: this.rttvar,
      rtt_ms: percentiles(this.rtts),
      queue_ms: percentiles(this.rtts.map((r) => r - minRtt)),
      ipdv_abs_ms: percentiles(this.ipdv.map(Math.abs)),
      up_excess_ms: this.minUp == null ? null : percentiles(this.ups.map((u) => u - this.minUp)),
      down_excess_ms: this.minDown == null ? null : percentiles(this.downs.map((d) => d - this.minDown)),
      // send side: what we asked the browser for vs what it did with it
      send_gap_ms: percentiles(this.sendGaps),
      write_ms: percentiles(this.writeMs),
      queue_depth: percentiles(this.queueDepth),
      timeline: binTimeline(this.records ?? [], this.firstSend ?? 0, ['rtt_ms', 'queue_ms', 'srtt_ms']),
    };
    if (keepRecords) { out.records = this.records; out.send_records = this.sendRecords; }
    return out;
  }
}

// ---- responder: reads the timestamps inside the stream ----------------------

export class RttStreamReceiver {
  constructor(keepRecords = true) {
    this.received = 0; this.duplicates = 0; this.reordered = 0;
    this.firstSeq = null; this.maxSeq = -1;
    this.jitter = 0; this.minOwd = null;
    this.records = keepRecords ? [] : null;
    this.seen = new Set();
    this.prev = null;
    this.owds = []; this.iat = []; this.sendIat = []; this.ipdv = [];
    this.firstRecv = null;
  }

  onPing(pkt, recvTs) {
    if (this.seen.has(pkt.seq)) this.duplicates++;
    this.seen.add(pkt.seq);
    this.received++;
    if (this.firstSeq == null) { this.firstSeq = pkt.seq; this.firstRecv = recvTs; }
    if (pkt.seq < this.maxSeq) this.reordered++;
    this.maxSeq = Math.max(this.maxSeq, pkt.seq);

    const owd = recvTs - pkt.sendTs;    // one-way delay + clock offset
    this.minOwd = this.minOwd == null ? owd : Math.min(this.minOwd, owd);
    let iat = null, sendIat = null, d = null;
    if (this.prev) {
      sendIat = pkt.sendTs - this.prev[0];
      iat = recvTs - this.prev[1];
      d = iat - sendIat;                // RFC 3550 D(i-1, i)
      this.jitter += (Math.abs(d) - this.jitter) * JITTER_GAIN;
      this.iat.push(iat); this.sendIat.push(sendIat); this.ipdv.push(d);
    }
    this.prev = [pkt.sendTs, recvTs];
    this.owds.push(owd);

    const rec = {
      seq: pkt.seq, t: recvTs, send_ts: pkt.sendTs, owd_ms: owd,
      owd_excess_ms: owd - this.minOwd, iat_ms: iat, send_iat_ms: sendIat,
      ipdv_ms: d, jitter_ms: this.jitter,
    };
    this.records?.push(rec);
    return rec;
  }

  summary(keepRecords = false) {
    if (!this.received) return null;
    const expected = this.maxSeq - (this.firstSeq ?? 0) + 1;
    const lost = Math.max(0, expected - this.seen.size);
    const minOwd = this.minOwd ?? 0;
    const out = {
      received: this.received, expected, lost,
      loss_pct: expected ? (100 * lost) / expected : null,
      reordered: this.reordered, duplicates: this.duplicates,
      jitter_ms: this.jitter,
      iat_ms: percentiles(this.iat),
      send_iat_ms: percentiles(this.sendIat),
      ipdv_abs_ms: percentiles(this.ipdv.map(Math.abs)),
      owd_excess_ms: percentiles(this.owds.map((o) => o - minOwd)),
      timeline: binTimeline(this.records ?? [], this.firstRecv ?? 0, ['owd_excess_ms', 'iat_ms', 'jitter_ms']),
    };
    if (keepRecords) out.records = this.records;
    return out;
  }
}

// ---- the send loop ----------------------------------------------------------

// One packet per tick. The tick source must be a worker (see tick-worker.js):
// main-thread timers are throttled to ~1/s in a background tab, which would
// stall the stream. Timer wake-ups are never exact, so the gaps we actually
// achieved are reported too - and since each packet carries the timestamp of
// the moment it was handed over, a ragged send schedule shows up on the
// receiver as `send_iat` and never gets mistaken for network jitter.
export async function runRttStream({ send, flow, nowMs, monitor, intervalMs, durationS, size, tick, isOpen, onSample }) {
  const start = nowMs();
  const end = start + durationS * 1000;
  let n = 0;
  while (isOpen()) {
    const t = nowMs();
    if (t >= end) break;
    monitor.onSend(n, t);          // records the gap we actually achieved
    send(proto.encodePing(flow, n, t, size), n);
    n++;
    onSample?.(n, t - start);
    await tick.next();
  }
  return { sent: n, interval_ms: intervalMs, size, send_gap_ms: percentiles(monitor.sendGaps) };
}
